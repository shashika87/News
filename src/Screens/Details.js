import React, {useEffect} from 'react';
import {
  View,
  Text,
  Image,
  StyleSheet,
  SafeAreaView,
  TouchableOpacity,
  ScrollView,
} from 'react-native';
import {connect} from 'react-redux';
import {
  fetchArticleSelected,
  fetchArticleSelectedNext,
  fetchArticleSelectedPrev,
  articleFavoriteGet,
  articleFavoriteAdd,
  articleFavoriteUpdate,
} from '../actions/newsActions';
import {nextIndex, prevIndex} from '../Utils/Utils';
import ThumbsupIcon from './Component/ThumbsupIcon';

const Details = (props) => {
  useEffect(() => {
    const {article} = props.route.params;
    articleIndex(article);
  }, []);

  const articleIndex = (article) => {
    props.dispatch(fetchArticleSelected(article));
    props.dispatch(articleFavoriteGet(article));
    const selectedIndex = props.articles.findIndex(
      (element) => element.title === article.title,
    );
    const selectedNextIndex = nextIndex(selectedIndex, props.articles.length);
    const selectedNextArticle = props.articles[selectedNextIndex];
    props.dispatch(fetchArticleSelectedNext(selectedNextArticle));
    const selectedPrevIndex = prevIndex(selectedIndex, props.articles.length);
    const selectedPrevArticle = props.articles[selectedPrevIndex];
    props.dispatch(fetchArticleSelectedPrev(selectedPrevArticle));
    //alert(props.navigation.state);
    props.navigation.setOptions({title: article.source.name});
  };

  let {selectedArticle, prevSelectedArticle, nextSelectedArticle} = props;
  selectedArticle = selectedArticle ? selectedArticle : props.route.params;
  const {updateArticles} = props.route.params;
  if (props.navigation) {
    props.navigation.setOptions({
      headerRight: () => (
        <View flexDirection="row">
          <TouchableOpacity
            style={{padding: 10}}
            onPress={() => {
              if (selectedArticle.favorite == null) {
                selectedArticle.favorite = true;
                props.dispatch(articleFavoriteAdd(selectedArticle));
              } else {
                selectedArticle.favorite = true;
                props.dispatch(articleFavoriteUpdate(selectedArticle));
              }
              props.dispatch(articleFavoriteGet(selectedArticle));
              updateArticles(selectedArticle);
            }}>
            <View style={{transform: [{scale: 0.8}]}}>
              <ThumbsupIcon
                fill={props.favorite ? '#157EFB' : 'none'}
                style={{width: 25, height: 25}}
              />
            </View>
          </TouchableOpacity>
          <TouchableOpacity
            style={{padding: 5, marginTop: 5}}
            onPress={() => {
              if (selectedArticle.favorite == null) {
                selectedArticle.favorite = false;
                props.dispatch(articleFavoriteAdd(selectedArticle));
              } else {
                selectedArticle.favorite = false;
                props.dispatch(articleFavoriteUpdate(selectedArticle));
              }
              props.dispatch(articleFavoriteGet(selectedArticle));
              updateArticles(selectedArticle);
            }}>
            <View style={{transform: [{rotate: '-180deg'}, {scale: 0.8}]}}>
              <ThumbsupIcon
                style={{width: 25, height: 25}}
                fill={
                  props.favorite == null || props.favorite == true
                    ? 'none'
                    : '#157EFB'
                }
              />
            </View>
          </TouchableOpacity>
        </View>
      ),
    });
  }
  return (
    <SafeAreaView style={styles.container}>
      <ScrollView>
        <View>
          {selectedArticle.urlToImage == null ? (
            <View
              style={{
                flex: 1,
                justifyContent: 'center',
                textAlign: 'center',
                alignItems: 'center',
                height: 180,
                width: 300,
              }}>
              <Text
                style={{
                  marginTop: 50,
                  textAlign: 'center',
                }}>
                No Preview
              </Text>
            </View>
          ) : (
            <Image
              style={({height: 180}, styles.image)}
              source={{uri: selectedArticle.urlToImage}}
            />
          )}
          <View style={{flex: 1, flexDirection: 'column'}}>
            <Text style={{fontSize: 20, margin: 10}}>
              {selectedArticle.title}
              {selectedArticle.author ? '(' + selectedArticle.author + ')' : ''}
            </Text>
            <Text style={{fontSize: 16, margin: 10}}>
              {selectedArticle.description}
            </Text>
            <Text style={{fontSize: 16, margin: 10}}>
              {selectedArticle.content}
            </Text>
          </View>
        </View>
      </ScrollView>
      <View style={styles.prevNext}>
        <TouchableOpacity
          style={styles.button}
          onPress={() => {
            articleIndex(prevSelectedArticle);
          }}>
          <View style={styles.TriangleShapeCSSPrevSide} />
        </TouchableOpacity>
        <TouchableOpacity
          style={styles.button}
          onPress={() => {
            articleIndex(nextSelectedArticle);
          }}>
          <View style={styles.TriangleShapeCSS} />
        </TouchableOpacity>
      </View>
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: 'white',
    width: '100%',
  },
  title: {
    textAlign: 'center',
    marginVertical: 8,
  },
  prevNext: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    flexDirection: 'row',
    alignContent: 'center',
    textAlign: 'center',
    justifyContent: 'space-between',
    backgroundColor: 'lightgray',
    width: '100%',
    padding: 20,
  },
  separator: {
    marginVertical: 8,
    borderBottomColor: '#737373',
    borderBottomWidth: StyleSheet.hairlineWidth,
  },
  TriangleShapeCSS: {
    width: 0,
    height: 0,
    borderLeftWidth: 10,
    borderRightWidth: 10,
    borderBottomWidth: 20,
    transform: [{rotate: '90deg'}],
    borderStyle: 'solid',
    backgroundColor: 'transparent',
    borderLeftColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: 'black',
  },

  TriangleShapeCSSPrevSide: {
    width: 0,
    height: 0,
    borderLeftWidth: 10,
    borderRightWidth: 10,
    borderBottomWidth: 20,
    transform: [{rotate: '-90deg'}],
    borderStyle: 'solid',
    backgroundColor: 'transparent',
    borderLeftColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: 'black',
  },

  SquareShapeCSS: {
    width: 20,
    height: 20,
    transform: [{rotate: '90deg'}],
    borderStyle: 'solid',
    backgroundColor: 'black',
    borderLeftColor: 'transparent',
    borderRightColor: 'transparent',
    borderBottomColor: '#00BCD4',
  },
  image: {
    height: 180,
    width: '100%',
    borderWidth: 1.0,
    borderColor: 'lightgray',
  },
  imageHolder: {
    margin: 5,
    height: 160,
    flex: 1,
    flexDirection: 'column',
    alignItems: 'center',
  },
});

const mapStateToProps = (state) => ({
  articles: state.news.articles,
  selectedArticle: state.news.selectedArticle,
  prevSelectedArticle: state.news.prevSelectedArticle,
  nextSelectedArticle: state.news.nextSelectedArticle,
  favorite: state.news.favorite,
});

export default connect(mapStateToProps)(Details);
