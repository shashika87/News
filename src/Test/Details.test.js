import React from 'react';
import Details from '../Screens/Details';
import {createStore, applyMiddleware} from 'redux';
import rootReducer from '../reducer/rootReducer';
import thunk from 'redux-thunk';
import {
  fetchNewsBegin,
  fetchNewsSuccess,
  fetchNewsailure,
} from '../actions/newsActions';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import items from './MockData/SearchResult';
import {TouchableOpacity, Image, View, TextInput} from 'react-native';
import {
  FETCH_NEWS_BEGIN,
  FETCH_NEWS_SUCCESS,
  FETCH_NEWS_FAILURE,
  FETCH_ARTICLE_SELECTED,
  FETCH_ARTICLE_SELECTED_PREV,
  FETCH_ARTICLE_SELECTED_NEXT,
  ARTICLE_FAVORITE_ADD,
  ARTICLE_FAVORITE_UPDATE,
  ARTICLE_FAVORITE_GET,
} from '../actions/newsActions';

const mockStore = configureStore([thunk]);

let store;
describe('Details', () => {
  let instance;
  var component;
  beforeEach(() => {
    // Simulate fetching a user from an API and loading it into state.
    var articles = {};
    store = mockStore({
      news: {
        searchText: '',
        items: [],
        loading: false,
        error: null,
        selectedArticle: items[1],
        prevSelectedArticle: items[0],
        nextSelectedArticle: items[2],
      },
    });

    //store.dispatch = jest.fn();
    let route = {params:{selectedArticle:items[1]}}
    component = renderer.create(
      <Provider store={store}>
        <Details route={route} />
      </Provider>,
    );
  });

  it('should render Details', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });
});
