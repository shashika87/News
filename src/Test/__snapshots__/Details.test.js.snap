// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Details should render Details 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
      "width": "100%",
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flex": 1,
            "flexDirection": "column",
            "height": 160,
            "margin": 5,
            "position": "relative",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "https://i.guim.co.uk/img/media/584082b5964a99b272e0373fff2728800437f7ba/795_224_3120_1872/master/3120.jpg?width=1200&height=630&quality=85&auto=format&fit=crop&overlay-align=bottom%2Cleft&overlay-width=100p&overlay-base64=L2ltZy9zdGF0aWMvb3ZlcmxheXMvdGctZGVmYXVsdC5wbmc&enable=upscale&s=1a60744b7c5f174356993242ad0317bf",
            }
          }
          style={
            Object {
              "borderColor": "lightgray",
              "borderWidth": 1,
              "height": "100%",
              "resizeMode": "cover",
              "width": "100%",
            }
          }
        />
        <View
          style={
            Object {
              "flexDirection": "column",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 20,
                "margin": 10,
              }
            }
          >
            Kelly Preston, actor and wife of John Travolta, dies aged 57 from breast cancer - The Guardian
            (Guardian staff reporter)
          </Text>
          <Text
            style={
              Object {
                "fontSize": 16,
                "margin": 10,
              }
            }
          >
            Travolta says ‘Kelly’s love and life will always be remembered’ after revealing actor’s death following two-year illness
          </Text>
          <Text
            style={
              Object {
                "fontSize": 16,
                "margin": 10,
              }
            }
          >
            Kelly Preston, the actor wife of John Travolta who appeared in movies such as Jerry Maguire and Gotti, has died from breast cancer aged 57.
Travolta said she had been sick for two years, and thanked… [+483 chars]
          </Text>
        </View>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "alignContent": "center",
        "backgroundColor": "lightgray",
        "bottom": 0,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "left": 0,
        "padding": 20,
        "position": "absolute",
        "right": 0,
        "textAlign": "center",
        "width": "100%",
      }
    }
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "borderBottomColor": "black",
            "borderBottomWidth": 20,
            "borderLeftColor": "transparent",
            "borderLeftWidth": 10,
            "borderRightColor": "transparent",
            "borderRightWidth": 10,
            "borderStyle": "solid",
            "height": 0,
            "transform": Array [
              Object {
                "rotate": "-90deg",
              },
            ],
            "width": 0,
          }
        }
      />
    </View>
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "transparent",
            "borderBottomColor": "black",
            "borderBottomWidth": 20,
            "borderLeftColor": "transparent",
            "borderLeftWidth": 10,
            "borderRightColor": "transparent",
            "borderRightWidth": 10,
            "borderStyle": "solid",
            "height": 0,
            "transform": Array [
              Object {
                "rotate": "90deg",
              },
            ],
            "width": 0,
          }
        }
      />
    </View>
  </View>
</RCTSafeAreaView>
`;
